int32 mode
int32 num_step
bool smoothing
bool save_pc
bool load_saved_pc